### PR TITLE
Rework keyboard handling to use keycodes (allowing for non-qwerty layouts)

### LINKED ISSUE #3177: Keyboard layout wrong in keyboard shortcuts
**OS:** MacOS 10.11.4
**Version:** 0.0.4

When settings keyboard shortcuts, if you have an AZERTY keyboard, it will display the QWERTY character.

For example, if you set a shortcut to A, it will be shown as Q. However it will work when pressing A.
- [ ] Reproducable in RCT2 (vanilla)?
- [ ] Multiplayer?

### PR DESCRIPTION
**Pull request by Goddesen**
This PR removes (most of the) usage of scancodes for keyboard input, allowing for keyboard layouts other than qwerty. This should fix #3177.

Current issues
- None?

The old code had a 256-long array of chars indexed by scancodes, where the value at subscript `i` would be `1` if scancode `i` had been pressed during this update. This has been reworked to be an array of `keypress` structs:

``` C
typedef struct {
    SDL_Keycode keycode;
    uint16 mod;
} keypress;
```

This allows for a wider range of possible shortcuts (media keys, meta/super key as a modifier on windows/linux) in addition to supporting different keyboard layouts, while staying relatively compact and platform agnostic.

Furthermore, modifier handling for shortcuts has been decoupled from `gInputPlaceObjectModifier` and the global per-update modifier state, instead favoring separate modifiers recorded for each keypress. This will let the user trigger shortcut combinations such as `C-s` and `S-k` in one update cycle without triggering potentially undefined behavior in the shortcut handling.

Shortcut keys that need to be held (map scrolling) and modifiers (used in object placement) now have their own separate minimal state. This is similar to the old system, but is now handled as part of keyboard input instead of separately getting the entire keyboard state.

An additional possible significant change: On _gain_ or _loss_ of window focus, all non-window events are ignored. This makes sure no input is handled if the window loses focus on in the same update as the keys were pressed. Allows e.g. binding TAB to _underground view_ without triggering the _underground view_ shortcut every time you _alt-TAB_ away from the game. Multiple window events per update are still handled, as I'm sure there would have been issues at some point where people would manage to lose and subsequently gain focus within one game update, thereby blocking inputs indefinitely.

This should now be ready for testing. @marijnvdwerf had problems where keyboard input will randomly not register _keycodes_ properly thus ignoring the input (though the _scancodes_ are fine). Please report if you have this issue and provide relevant output given by [this branch](https://github.com/Goddesen/OpenRCT2/tree/3908_debug).

### CODE REVIEW

File: src/platform/platform.h [hunk 1]
```diff
@@ -130,11 +140,18 @@ void platform_toggle_windowed_mode();
...
 void platform_stop_text_input();
 void platform_get_date(rct2_date *out_date);
 void platform_get_time(rct2_time *out_time);
+bool platform_keypress_equals(keypress k1, keypress k2);
+bool platform_keypress_in_update(keypress k);
+bool platform_check_alt(void);
+bool platform_check_ctrl(void);
+bool platform_check_gui(void);
+bool platform_check_shift(void);
+#define CHECK_PLACE_OBJECT_MOD_COPY_Z platform_check_ctrl()
```
Comment (duncanspumpkin) [inline src/platform/platform.h]: Can you just use the actual function call instead of this macro. I don't think the macro really adds anything.

Comment (Goddesen) [inline src/platform/platform.h]: Can do.

File: src/input.c [hunk 2]
```diff
@@ -1658,50 +1619,15 @@ void game_handle_key_scroll()
...
-			scrollY = 1;
-			break;
-		case SHORTCUT_SCROLL_MAP_RIGHT:
-			scrollX = 1;
-			break;
-		}
-	}
+	if (gKeysHeld[1])
+		scrollX = -1;
+	else if (gKeysHeld[3])
```
Comment (Goddesen) [inline src/input.c]: Currently scrolling up and left has precedence over down and right, respectively. No real reason.
I can make either 1) directions cancel each other out, or 2) always use most recently pressed key. The latter might require some extra state.

Comment (IntelOrca) [inline src/input.c]: If you make it use the last pressed key, then you would have to also make sure it switches back if you let go of that last pressed key.

Comment (Goddesen) [inline src/input.c]: Certainly doable. Is this really something people would (a)buse enough to warrant introducing more complex state?

Comment (Goddesen) [inline src/input.c]: On second thought it needn't be that complex. I'll make something happen.

**Comment by Goddesen** [comment]
If someone with a true N-key rollover keyboard could test map scrolling that'd be great. :)

**Comment by danidoedel** [comment]
No issues with map scrolling for me ;)
Other shortcuts like zooming, rotating, opening a window etc. only work after pressing the button a few times though.

**Comment by Goddesen** [comment]
I think there might be a loop running wild somewhere, I'll look into that.

**Comment by Goddesen** [comment]
There was an issue with (I think) an infinite loop earlier, but that might have been connected to the faulty object loading code from earlier. The only thing that remains should be to figure out why @marijnvdwerf has weird issues.

**Comment by danidoedel** [comment]
Still doesn't detect every keystroke. No issues with map scrolling...

**Comment by Goddesen** [comment]
@danidoedel Could you please try out the branch I link to in the end? This has verbose output. Please provide some sample output along with which keystrokes did and did not trigger.

**Comment by danidoedel** [comment]
@Goddesen I don't have any issues while using your debug branch... That's weird.

**Comment by IntelOrca** [comment]
@Goddesen what is the status of this PR?

**Comment by Nubbie** [comment]
Might fix #1164

**Comment by Goddesen** [comment]
I have experienced the issue with keys not registering myself, but have for weeks now been unable to reproduce it. I'm stumped...

**Comment by IntelOrca** [comment]
@Goddesen could you rebase this?

**Comment by Goddesen** [comment]
This is now rebased and up to date, provided that integration tests pass. I have not yet noticed previously mentioned Mac-specific problems but it would be nice if a bunch of people could test this :)

**Comment by janisozaur** [comment]
It survived a furious-keyboard-mashing test, that's a good start 👍

File: src/rct2.c [hunk 3]
```diff
@@ -0,0 +1,515 @@
+#pragma region Copyright (c) 2014-2016 OpenRCT2 Developers
```
Comment (janisozaur) [inline src/rct2.c]: Looks like an old version of this file that shouldn't be here?

**Comment by Goddesen** [comment]
Removed erroneously added old file `src/rct2.c` and rebased again.

**Comment by danidoedel** [comment]
My issue mentioned above still exists, but it works just fine on my laptop.

**Comment by Goddesen** [comment]
@danidoedel OS and system specs? Maybe I can try to recreate the issues you're having.

**Comment by danidoedel** [comment]
Using Windows 10 x64 with German system language if that matters. What other system specs do you want to know?

**Comment by Goddesen** [comment]
I was thinking maybe if your computer was abnormally weak, it might affect some weird parts of the code. But maybe not. I'll give this a try on win10 when I get the chance 👍

**Comment by Gymnasiast** [comment]
@Goddesen I'd also try setting your keyboard to a German layout when you test on W10.

**Comment by Goddesen** [comment]
After finally setting up my build environment on Windows 10 I can confirm that there are issues. Gonna try to look into them. Slowly, probably :)

**Comment by Goddesen** [comment]
The issue with disappearing keystrokes should now be gone! :)

**Comment by danidoedel** [comment]
Yeah works fine now! Good job :)

**Comment by marijnvdwerf** [comment]
The console still opens using the key next to <kbd>1</kbd> for me

File: src/openrct2/input.c [hunk 4]
```diff
@@ -1463,86 +1445,76 @@ void title_handle_keyboard_input()
...
+
+	gNumKeysPressed = 0;
 }
 
 /**
  *
  *  rct2: 0x00406CD2
  */
-sint32 get_next_key()
+sint32 get_next_key(keypress *kp)
```
Comment (marijnvdwerf) [inline src/openrct2/input.c]: isn't this just a bool?

**Comment by janisozaur** [comment]
@marijnvdwerf is this not correct? What should it open with?

File: src/openrct2/platform/shared.c [hunk 5]
```diff
@@ -554,10 +580,73 @@ void platform_process_messages()
...
+
+	if (key.keycode == SDLK_LEFT) {
+		textinputbuffer_cursor_left(&gTextInput);
+		console_refresh_caret();
+
+	} else if (key.keycode == SDLK_RIGHT) {
+		textinputbuffer_cursor_right(&gTextInput);
+		console_refresh_caret();
+
+	} else if (platform_keypress_equals(key, (keypress){ SDLK_v, KEYBOARD_PRIMARY_MODIFIER })) {
```
Comment (marijnvdwerf) [inline src/openrct2/platform/shared.c]: Might be a dump question, but do all platforms only allow <kbd>cmd/ctrl</kbd>+<kbd>v</kbd> as paste shortcut?

**Comment by marijnvdwerf** [comment]
On English keyboards it opens with the backtick key. Which is situated between <kbd>shift</kbd>and <kbd>z</kbd> for me. Also, opening the console types a character, but that's an existing issue.

File: src/openrct2/platform/platform.h [hunk 6]
```diff
@@ -116,10 +122,19 @@ typedef struct file_dialog_desc {
...
+	SDL_Keycode keycode;
+	uint16 mod;
+} keypress;
+
 extern openrct2_cursor gCursorState;
-extern const uint8 *gKeysState;
-extern uint8 *gKeysPressed;
-extern uint32 gLastKeyPressed;
+extern keypress *gKeysPressed;
+extern int gMapKeysStackHorizontal[2]; // State for map scroll shortcut keys
```
Comment (Goddesen) [inline src/openrct2/platform/platform.h]: Even when it's an enum?

File: src/openrct2/input.c (hunk 4 above)
Comment (Goddesen) [inline src/openrct2/input.c]: My C89 is showing through! 😲

**Comment by Goddesen** [comment]
@marijnvdwerf US International keyboards have backtick on the key next to 1 :).

**Comment by Goddesen** [comment]
I'll look into how to handle dead keys properly (if at all).

**Comment by marijnvdwerf** [comment]
I'm aware. But shouldn't this PR actually fix that?

**Comment by Goddesen** [comment]
Fix it? Fix what?

**Comment by marijnvdwerf** [comment]
Fix the backtick being detected by the scancode of the key, as opposed to the actual symbol the key produces.

**Comment by Goddesen** [comment]
In this PR it is detected by the backtick keycode: https://github.com/OpenRCT2/OpenRCT2/pull/3908/files#diff-974543acd820d663a684cbe24055cfb3R1414

**Comment by Gymnasiast** [comment]
Can you make it check for both backtick and dead backtick?

**Comment by Goddesen** [comment]
@Gymnasiast Technically yes, but I need to look into how that would handle across keyboard layouts. Will check into this again shortly, after my exams :)

**Comment by Gymnasiast** [comment]
@AaronVanGeffen and I discussed this.

So many changes have occurred in the meantime that rebasing is probably not worth the effort. For this reason we'll close this. If anyone feels like picking it up, please feel free to do so!

**Pull request closed without merging**